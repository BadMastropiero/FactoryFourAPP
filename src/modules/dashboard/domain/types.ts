import { z } from 'zod';

import { serviceStatusSchema } from './schemas';

export type ServiceStatus = z.infer<typeof serviceStatusSchema>;

export type ExtendedServiceStatus = ServiceStatus & {
  title: string;
};
