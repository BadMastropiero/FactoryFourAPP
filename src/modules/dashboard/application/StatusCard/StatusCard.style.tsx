import styled, { css } from 'styled-components';

export const StyledTitle = styled.div`
  font-size: ${(props) => props.theme.fontSizes.l};
  font-weight: 700;
  text-transform: uppercase;
  text-align: center;
`;

export const StyledStatus = styled.div<{ success: boolean }>`
  font-size: ${(props) => props.theme.fontSizes.s};
  color: ${(props) =>
    props.success
      ? props.theme.colors.success.contrast
      : props.theme.colors.error.contrast};
  background-color: ${(props) =>
    props.success ? props.theme.colors.success.base : props.theme.colors.error.base};
  font-weight: 700;
  text-transform: uppercase;
  text-align: center;
  border-radius: ${(props) => props.theme.borderRadius.base};
  width: 100px;
  transition: width 0.2s ${(props) => props.theme.easings.wiggle};
  padding: calc(${(p) => p.theme.layout.vGap} / 2) calc(${(p) => p.theme.layout.hGap} / 1);
`;

export const StyledTime = styled.time``;

export const StyledErrorMessage = styled.div`
  font-size: ${(props) => props.theme.fontSizes.s};
  color: ${(props) => props.theme.colors.error.base};
  text-transform: uppercase;
  font-weight: 700;
`;

export const StyledHostname = styled.div``;

export const StyledCard = styled.div<{ selected: boolean }>`
  display: flex;
  align-items: center;
  flex-direction: column;
  gap: calc(${(props) => props.theme.layout.vGap} / 2);
  height: 250px;
  background-color: ${(props) => props.theme.colors.lighter.base};
  padding: ${(props) => [props.theme.layout.vGap, props.theme.layout.hGap].join(' ')};
  transition:
    width 0.2s ${(props) => props.theme.easings.wiggle},
    box-shadow 0.1s ${(props) => props.theme.easings.base},
    transform 0.1s ${(props) => props.theme.easings.base};
  /* border: 1px solid ${(props) => props.theme.colors.primary.base}; */
  border-radius: ${(props) => props.theme.borderRadius.base};
  box-shadow: ${(props) => props.theme.shadows.base};

  &:hover {
    box-shadow: ${(props) => props.theme.shadows.hover};
    ${StyledStatus} {
      /* transform: scale(1.05); */
      width: ${(p) => !p.selected && '100%'};
    }
  }

  ${(p) =>
    p.selected &&
    css`
      position: absolute;
      top: 0;
      left: 0;
      right: 0;
      bottom: 0;
      width: 100%;
      margin: auto;
      min-height: 80vh;
      max-width: ${p.theme.layout.maxWidth};
    `}
`;
