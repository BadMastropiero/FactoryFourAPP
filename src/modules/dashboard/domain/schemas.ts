import { z } from 'zod';

export const serviceStatusSchema = z.object({
  success: z.boolean(),
  message: z.string(),
  hostname: z.string(),
  time: z.number().optional(),
});
