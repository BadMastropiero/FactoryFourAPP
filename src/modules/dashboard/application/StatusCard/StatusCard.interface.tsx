import { ExtendedServiceStatus, ServiceStatus } from '../../domain/types';

export interface StatusCardProps {
  item: ExtendedServiceStatus;
  history?: ServiceStatus[];
}
