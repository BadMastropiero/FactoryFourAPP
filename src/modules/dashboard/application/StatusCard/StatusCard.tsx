import { useRef, useState } from 'react';

import useLockBodyScroll from '../../../../hooks/useLockBodyScroll';
import useOnClickOutside from '../../../../hooks/useOnClickOutside';
import { StatusCardProps } from './StatusCard.interface';
import {
  StyledCard,
  StyledErrorMessage,
  StyledHistoryItem,
  StyledHistoryList,
  StyledHostname,
  StyledStatus,
  StyledTime,
  StyledTitle,
} from './StatusCard.style';

export default function StatusCard({ item, history }: StatusCardProps) {
  const [selected, setSelected] = useState<boolean>(false);
  const ref = useRef(null);
  useOnClickOutside(ref, (e) => {
    setSelected(false);
    e.stopPropagation();
    // e.stopImmediatePropagation();
  });
  useLockBodyScroll(selected, [selected]);

  return (
    <>
      {selected && <br />}
      <StyledCard
        data-testid="styled-card"
        onClick={(e) => {
          setSelected(true);
          e.stopPropagation();
        }}
        selected={selected}
        ref={ref}
      >
        {!selected && (
          <>
            <StyledTitle>{item?.title}</StyledTitle>
            <StyledStatus success={item?.success}>
              {item?.success ? 'Healthy' : 'Error'}
            </StyledStatus>
            <StyledTime>
              {item?.time ? new Date(item.time).toLocaleTimeString() : undefined}
            </StyledTime>
            {!item?.success && (
              <StyledErrorMessage>
                <p>Outage</p>
                <p>{item.message}</p>
              </StyledErrorMessage>
            )}
            <StyledHostname>{item?.hostname}</StyledHostname>
          </>
        )}
        {selected && (
          <>
            <StyledTitle>History</StyledTitle>
            <StyledHistoryList>
              {history?.map((item, i) => (
                <StyledHistoryItem key={i}>
                  <StyledStatus success={item?.success}>
                    {item?.success ? 'Healthy' : 'Error'}
                  </StyledStatus>
                  <b>Time:</b>
                  <StyledTime>
                    {item?.time ? new Date(item.time).toLocaleTimeString() : undefined}
                  </StyledTime>
                  {!item?.success && (
                    <>
                      <b>Details:</b>
                      <StyledErrorMessage>
                        <p>Outage {item.message}</p>
                      </StyledErrorMessage>
                    </>
                  )}
                  {item?.hostname && (
                    <>
                      <b>Host:</b>
                      <StyledHostname>{item?.hostname}</StyledHostname>
                    </>
                  )}
                </StyledHistoryItem>
              ))}
            </StyledHistoryList>
          </>
        )}
      </StyledCard>
    </>
  );
}
