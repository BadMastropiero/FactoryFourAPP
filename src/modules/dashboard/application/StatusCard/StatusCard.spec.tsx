import '@testing-library/jest-dom';

import { fireEvent, render, screen } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { describe, expect, it } from 'vitest';

import { theme } from '../../../../styles/theme';
import { ExtendedServiceStatus } from '../../domain/types';
import StatusCard from './StatusCard';

const mockStatus: ExtendedServiceStatus = {
  title: 'Server 1',
  success: true,
  time: 121212,
  hostname: 'localhost',
  message: '',
};

describe('StatusCard', () => {
  it('renders correctly', () => {
    const { asFragment } = render(
      <ThemeProvider theme={theme}>
        <StatusCard item={mockStatus} />
      </ThemeProvider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders with healthy status', () => {
    render(
      <ThemeProvider theme={theme}>
        <StatusCard item={{ ...mockStatus, success: true }} />
      </ThemeProvider>,
    );
    expect(screen.getByText('Healthy')).toBeInTheDocument();
    expect(screen.queryByText('Outage')).not.toBeInTheDocument();
  });

  it('renders with error status', () => {
    render(
      <ThemeProvider theme={theme}>
        <StatusCard item={{ ...mockStatus, success: false }} />
      </ThemeProvider>,
    );
    expect(screen.getByText('Error')).toBeInTheDocument();
    expect(screen.getByText('Outage')).toBeInTheDocument();
  });

  it('toggles selected state on click', () => {
    render(
      <ThemeProvider theme={theme}>
        <StatusCard item={mockStatus} />
      </ThemeProvider>,
    );
    const card = screen.getByTestId('styled-card');
    fireEvent.click(card);
  });
});
