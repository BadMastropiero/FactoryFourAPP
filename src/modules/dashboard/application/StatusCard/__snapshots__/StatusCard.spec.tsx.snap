// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StatusCard > renders correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-jSguLX etPCaA"
    data-testid="styled-card"
  >
    <div
      class="sc-bdfCDU cFAzbI"
    >
      Server 1
    </div>
    <div
      class="sc-gsTDqH fiKUpR"
    >
      Healthy
    </div>
    <time
      class="sc-dlfmHC"
    >
      6:02:01 PM
    </time>
    <div
      class="sc-eCstZk"
    >
      localhost
    </div>
  </div>
</DocumentFragment>
`;
